### PR TITLE
[dynamo] canonicalize output_graph node order (with config flag)

### PR DESCRIPTION
**Pull request by williamwen42**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* #181781
* #181780
* #181779
* #181778
* #181777
* #181776
* __->__ #181775
* #187111
* #187100

cc @albanD @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @jataylo @azahed98

### CODE REVIEW

File: torch/_dynamo/config.py [hunk 1]
```diff
@@ -910,6 +910,11 @@ def default_debug_dir_root() -> str:
 # - False: never clear regardless of backend
 invalidate_compile_context_weakrefs: bool | None = None
 
+# Reorder and rename output graph nodes into a canonical topological order so
+# that structurally equivalent graphs (e.g., same model traced with different
+# dict iteration orders across distributed ranks) produce identical FX graphs.
+canonicalize_output_graph_node_order = False
```
Comment (aorenste) [inline torch/_dynamo/config.py]:
```suggestion
canonicalize_output_graph_node_order: bool = False
```

File: torch/_dynamo/output_graph.py [hunk 2]
```diff
@@ -492,6 +493,161 @@ class ExportMetaData:
...
+    if node.op == "call_method":
+        # Node.is_impure() returns False for all call_method nodes, but
+        # in-place methods (e.g. .requires_grad_(), .copy_()) mutate state.
+        return not node.target.endswith("_")  # pyrefly: ignore[missing-attribute]
+    if node.op != "call_function":
+        return True
+    # Node.is_impure() covers: mutable OpOverload schemas, effects system,
+    # _side_effectful_functions, and random ops.
+    if node.is_impure():
+        return False
```
Comment (aorenste) [inline torch/_dynamo/output_graph.py]: Shouldn't this check be first?

File: torch/_dynamo/output_graph.py [hunk 3]
```diff
@@ -492,6 +493,161 @@ class ExportMetaData:
...
+    # 2. out= kwarg with a Node value indicates mutation (e.g. torch.add(y, 1, out=x)).
+    # 3. Targets with no FX Node arguments are likely state-changing operations
+    #    (e.g., _vmap_increment_nesting, _set_fwd_grad_enabled).
+    if not isinstance(node.target, torch._ops.OpOverload):
+        name = getattr(node.target, "__name__", "")
+        if name.endswith("_"):
+            return False
+        if isinstance(node.kwargs.get("out"), fx.Node):
+            return False
+        return bool(node.all_input_nodes)
```
Comment (aorenste) [inline torch/_dynamo/output_graph.py]: This will categorize constant ops as unsafe. Is that intended? Maybe we should have some kind of list?

File: torch/_dynamo/output_graph.py [hunk 4]
```diff
@@ -492,6 +493,161 @@ class ExportMetaData:
...
+
+    Operates in-place on the graph to preserve node object identity (required
+    by GraphRegionTracker and other infrastructure that holds node references).
+    """
+    indeg: dict[fx.Node, int] = dict.fromkeys(graph.nodes, 0)
+
+    # Build standard data-dependency edges
+    for node in graph.nodes:
+        for user in node.users:
+            indeg[user] += 1
```
Comment (aorenste) [inline torch/_dynamo/output_graph.py]: Isn't this just:
```
indeg = {node: len(node.all_input_nodes) for node in graph.nodes}
```

File: torch/_dynamo/output_graph.py [hunk 5]
```diff
@@ -492,6 +493,161 @@ class ExportMetaData:
...
+    extra_users: dict[fx.Node, list[fx.Node]] = collections.defaultdict(list)
+    prev_barrier: fx.Node | None = None
+    segment_reorderable: list[fx.Node] = []
+    for node in graph.nodes:
+        if node.op in ("placeholder", "get_attr", "output"):
+            continue
+        if not _is_safe_to_reorder(node):
+            if prev_barrier is not None:
+                extra_users[prev_barrier].append(node)
+                indeg[node] += 1
```
Comment (aorenste) [inline torch/_dynamo/output_graph.py]: This is the same on both arms of the `if` - move it out to dedup.

File: torch/_dynamo/output_graph.py [hunk 6]
```diff
@@ -492,6 +493,161 @@ class ExportMetaData:
...
+            return (3,)
+        else:
+            input_indices = tuple(canonical_idx[n] for n in node.all_input_nodes)
+            return (2, input_indices)
+
+    # Seed the heap with nodes that have no dependencies
+    ready: list[tuple[tuple[Any, ...], int, fx.Node]] = []
+    for node in graph.nodes:
+        if indeg[node] == 0:
+            heapq.heappush(ready, (_canonical_key(node), counter, node))
```
Comment (aorenste) [inline torch/_dynamo/output_graph.py]: This is the main reason for requesting changes:
I think this counter is a problem (here and all uses below). It's going to cause pure functions with identical inputs to not be reordered:
```
add(a, b)
mul(a, b)
```
vs
```
mul(a, b)
add(a, b)
```
it seems like you want them to be reordered to be consistent but the counter will prevent that. I think you want to remove the counter and instead add the target to the key as the tie-breaker? By the time you get to the heappush if two keys are the same the order doesn't matter because they should be indistinguishable (one could argue we should de-dup them at that point).

**Review by aorenste — CHANGES_REQUESTED** [review]

File: torch/_dynamo/output_graph.py [hunk 7]
```diff
@@ -492,6 +493,161 @@ class ExportMetaData:
...
+        name = getattr(node.target, "__name__", "")
+        if name.endswith("_"):
+            return False
+        if isinstance(node.kwargs.get("out"), fx.Node):
+            return False
+        return bool(node.all_input_nodes)
+    return True
+
+
+def _canonicalize_graph(graph: fx.Graph) -> fx.Graph:
```
Comment (aorenste) [inline torch/_dynamo/output_graph.py]: This functionality isn't specific to dynamo, is it? Maybe put it in torch/fx/passes?

File: torch/_dynamo/output_graph.py [hunk 8]
```diff
@@ -492,6 +493,161 @@ class ExportMetaData:
...
+    # FX Graph.create_node uses for auto-generated names).
+    from torch.fx.graph import _Namespace
+
+    ns = _Namespace()
+    for node in graph.nodes:
+        candidate = graph._target_to_str(node.target)
+        new_name = ns.create_name(candidate, node)
+        node.name = new_name
+    # Replace the graph's namespace so future node creation stays consistent
+    graph._graph_namespace = ns
```
Comment (aorenste) [inline torch/_dynamo/output_graph.py]: This "rename pass" seems useful on its own. We should put it as its own separate pass in torch/fx/passes.

File: torch/_dynamo/output_graph.py [hunk 9]
```diff
@@ -542,6 +542,99 @@ class ExportMetaData:
...
+        else:
+            source_name = ""
+        return (0, source_name)
+    elif node.op == "get_attr":
+        return (1, str(node.target))
+    elif node.op == "output":
+        return (3,)
+    else:
+        input_indices = tuple(canonical_idx[n] for n in node.all_input_nodes)
+        return (2, str(node.target), input_indices)
```
Comment (aorenste) [inline torch/_dynamo/output_graph.py]: (Claude Review) [question] `str(node.target)` embeds the target's `id()` for plain-Python-function targets, which is process-specific and works against the cross-rank determinism this pass exists to provide. Dynamo output graphs (pre-AOT) routinely carry such targets: `F.relu`/`F.softmax` stay as plain functions (`str()` -> `"<function relu at 0x...>"`) and `torch.relu` is a type-bound builtin (`"<built-in method relu of type object at 0x...>"`) -- both contain an address. The order only flips when two ready nodes tie on the `(2,)` prefix + `input_indices` and share a `__name__` (distinct same-named targets, or CSE duplicates of the same target), so the practical trigger is narrow -- but it's free to harden and the same-process tests can't catch it (the address is constant within one process). The `rename_nodes_to_canonical` pass right below already uses the address-free `graph._target_to_str(node.target)`; using that here too removes the `id()`.

File: torch/fx/passes/canonicalize.py [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+        else:
+            segment_reorderable.append(node)
+
+    canonical_idx: dict[fx.Node, int] = {}
+
+    # The counter is a tiebreaker that prevents heapq from comparing
+    # fx.Node objects (which have no __lt__). It only affects nodes with
+    # identical canonical keys -- i.e., structurally equivalent operations
+    # (same target, same input indices). Those are CSE candidates and
+    # genuinely interchangeable, so any ordering between them is canonical.
```
Comment (aorenste) [inline torch/fx/passes/canonicalize.py:100]: (Claude Review) [question] "any ordering between them is canonical" holds for graph *semantics* but not for the byte-identical names/edges this feature relies on. Two nodes with identical canonical keys (same target + same `input_indices`) get their `canonical_idx` decided by `counter`, i.e. original insertion order, which differs across ranks. If the duplicates have *distinguishable* users, that differing `canonical_idx` propagates into the users' keys and the graphs diverge. Concretely, `mul(x + x, 2) + neg(x + x)` produces two `add(x, x)` nodes, one feeding `mul` and one feeding `neg`: one rank emits `mul(add, ...)` / `neg(add_1)`, the other `mul(add_1, ...)` / `neg(add)` -- isomorphic but not name-identical, so a name-based hash still diverges. The earlier review thread's suggestion to dedup identical-key nodes would close this; otherwise the comment overstates the guarantee.

File: torch/_dynamo/output_graph.py [hunk 11]
```diff
@@ -542,6 +542,99 @@ class ExportMetaData:
...
+def _is_safe_to_reorder(node: fx.Node) -> bool:
+    """Check if a node is safe to reorder during graph canonicalization.
+
+    Builds on Node.is_impure() (used by DCE) with two additional checks for
+    cases it doesn't cover: in-place call_method nodes and non-OpOverload
+    state-changing functions detected by a no-node-arguments heuristic.
+    """
+    if node.op == "call_method":
+        return not node.target.endswith("_")  # pyrefly: ignore[missing-attribute]
+    if node.op != "call_function":
```
Comment (aorenste) [inline torch/_dynamo/output_graph.py:576]: (Claude Review) [cleanup] `call_module` falls through to `return True` here and is always treated as reorderable, bypassing `node.is_impure()` -- which for `call_module` consults the submodule's `_is_impure`. The docstring says this "builds on Node.is_impure()", but it does not for this op. This is not a live miscompile today (default `torch.compile` inlines all nn.Modules so no `call_module` nodes are produced, and the modules reachable under export never set `_is_impure`), but routing `call_module` through `node.is_impure()` -- or just treating it as a barrier -- is free insurance and makes the code match its docstring. Note this same plain-Python-function impurity gap is exactly why the call site has to special-case `in_compiled_autograd_region`.

File: torch/_dynamo/output_graph.py [hunk 12]
```diff
@@ -542,6 +542,99 @@ class ExportMetaData:
...
+        # _set_fwd_grad_enabled). This is intentionally conservative:
+        # pure constant-producing ops would also be treated as barriers,
+        # but those are rare in Dynamo output graphs (constants are
+        # typically lifted as placeholders or get_attr nodes).
+        if not node.all_input_nodes:
+            return False
+        # functorch batch dim ops modify the vmap interpreter stack.
+        if name in ("_add_batch_dim", "_remove_batch_dim"):
+            return False
+        return True
```
Comment (aorenste) [inline torch/_dynamo/output_graph.py]: nit: could just drop through here. Not sure if that's cleaner or not...

File: torch/_dynamo/output_graph.py [hunk 13]
```diff
@@ -542,6 +542,99 @@ class ExportMetaData:
...
+    return True
+
+
+def _canonical_key(node: fx.Node, canonical_idx: dict[fx.Node, int]) -> tuple[Any, ...]:
```
Comment (aorenste) [inline torch/_dynamo/output_graph.py]: Seems like the cleaner return type would be:
```
-> tuple[int, str] | tuple[int] | tuple[int, str, tuple[int, ...]]
```
Maybe it's moot because it's just feeding into a key - in which case maybe just make it `-> object`?

File: torch/_dynamo/output_graph.py [hunk 14]
```diff
@@ -2791,6 +2884,12 @@ def compile_and_call_fx_graph(
             # free a bit of memory
             self.real_value_cache.clear()
 
+            if (
+                config.canonicalize_output_graph_node_order
+                and not torch._dynamo.compiled_autograd.in_compiled_autograd_region
```
Comment (aorenste) [inline torch/_dynamo/output_graph.py:2893]: This guard needs a comment -- it's not obvious why compiled autograd is excluded. As far as I can tell: CA backward graphs contain side-effecting ops (`call_accumulate_grad` / `call_hook` / `call_backward`) whose plain-Python-function targets `is_impure()` reports as pure, so the barrier scheme would happily reorder them; and CA graphs use a fixed positional boxed placeholder layout (`inputs`/`sizes`/`scalars`/`hooks`/`packed_data`) that must not be reordered or renamed. A one-liner here would save the next reader the dig.

File: torch/fx/passes/canonicalize.py [hunk 15]
```diff
@@ -0,0 +1,142 @@
...
+    counter = 0
+    ready: list[tuple[tuple[Any, ...], int, fx.Node]] = []
+    for node in graph.nodes:
+        if indeg[node] == 0:
+            heapq.heappush(
```
Comment (aorenste) [inline torch/fx/passes/canonicalize.py]: nit (because the win is small):
It would be better to just heapq.append(...) and then heapq.heapify(ready) after the loop - O(n) instead of O(n log n). But presumably n is small so kind of moot.

File: torch/fx/passes/canonicalize.py [hunk 16]
```diff
@@ -0,0 +1,142 @@
...
+        for user in itertools.chain(cur.users, extra_users.get(cur, ())):
+            indeg[user] -= 1
+            if indeg[user] == 0:
+                heapq.heappush(
+                    ready,
+                    (canonical_key_fn(user, canonical_idx), counter, user),
+                )
+                counter += 1
+
+    if len(canonical_order) != len(list(graph.nodes)):
```
Comment (aorenste) [inline torch/fx/passes/canonicalize.py]: Unnecessary list - turns O(1) into O(n)
```suggestion
    if len(canonical_order) != len(graph.nodes):
```

**Review by aorenste — APPROVED** [review]
